File: androidApp/src/main/java/com/alifwyaa/azanmunich/android/ui/screens/home/HomeViewModel.kt
package com.alifwyaa.azanmunich.android.ui.screens.home

import androidx.compose.runtime.Stable
import androidx.lifecycle.viewModelScope
import com.alifwyaa.azanmunich.android.ui.screens.common.BaseState
import com.alifwyaa.azanmunich.android.ui.screens.common.BaseViewModel
import com.alifwyaa.azanmunich.android.ui.screens.common.DataPlaceHolder
import com.alifwyaa.azanmunich.domain.SharedStrings
import com.alifwyaa.azanmunich.domain.internal.platform.Screen
import com.alifwyaa.azanmunich.domain.internal.platform.SharedTrackingService
import com.alifwyaa.azanmunich.domain.model.SharedAzanModel
import com.alifwyaa.azanmunich.domain.model.SharedDateModel
import com.alifwyaa.azanmunich.domain.model.SharedResult
import com.alifwyaa.azanmunich.domain.model.SharedTimeUntil
import com.alifwyaa.azanmunich.domain.services.SharedAzanService
import com.alifwyaa.azanmunich.domain.services.SharedDateTimeService
import com.alifwyaa.azanmunich.domain.services.SharedLocalizationService
import com.alifwyaa.azanmunich.domain.services.SharedSettingsService
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.ensureActive
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

/**
 * HomeViewModel
 */
class HomeViewModel(
    private val azanService: SharedAzanService,
    private val dateTimeService: SharedDateTimeService,
    private val localizationService: SharedLocalizationService,
    private val settingsService: SharedSettingsService,
    trackingService: SharedTrackingService,
) : BaseViewModel<HomeViewModel.State>() {

    override val initialState: State by lazy {
        State(
            azanDate = State.AzanDate(
                date = dateTimeService.getToday(),
                westernFormattedDate = dateTimeService
                    .getWesternFormattedDate(dateTimeService.getToday()),
                islamicFormattedDate = dateTimeService
                    .getIslamicFormattedDate(dateTimeService.getToday())
            ),
        )
    }

    private val strings: SharedStrings
        get() = localizationService.strings

    private var lastFetchedNextAzan: SharedAzanModel? = null

    private var fetchNextAzanJob: Job? = null

    init {
        trackingService.trackScreen(Screen.HOME)
        scheduleNextAzanFetch()
        loadAzanList(dateTimeService.getToday())
        listenToSettingsChanges()
    }

    //region Public Methods

    /**
     * Retry using the last date
     */
    fun reloadAzanList() {
        loadAzanList(state.azanDate.date)
    }

    /**
     * Go to previous day
     */
    fun goToDayBefore() {
        val dayBefore = dateTimeService.getDayBefore(state.azanDate.date)
        loadAzanList(dayBefore)
    }

    /**
     * Go to next day
     */
    fun goToDayAfter() {
        val dayAfter = dateTimeService.getDayAfter(state.azanDate.date)
        loadAzanList(dayAfter)
    }

    /**
     * Go to next azan day
     */
    fun goToNextAzanDay() {
        val nextAzan = state.nextAzan

        if (nextAzan !is DataPlaceHolder.Success
            || nextAzan.data.dateModel == state.azanDate.date
        ) return

        loadAzanList(nextAzan.data.dateModel)
    }

    //endregion

    //region Helpers

    private fun loadAzanList(date: SharedDateModel) {
        computeAndUpdateState {
            it.copy(
                azanDate = State.AzanDate(
                    date = date,
                    westernFormattedDate = dateTimeService.getWesternFormattedDate(date),
                    islamicFormattedDate = dateTimeService.getIslamicFormattedDate(date),
                ),
                azanList = DataPlaceHolder.Loading
            )
        }

        viewModelScope.launch(Dispatchers.Main) {
            delay(300)
            val result = azanService.getAzanList(date)
                .mapSuccess { list ->
                    list.map { sharedAzanModel ->
                        State.Azan(
                            model = sharedAzanModel,
                            isNextAzan = sharedAzanModel.id == state.nextAzan.getOrNull()?.id,
                            isNotificationEnabled = settingsService.isAzanEnabled(sharedAzanModel.azanType)
                        )
                    }
                }
            when (result) {
                is SharedResult.Success -> {
                    computeAndUpdateState {
                        it.copy(azanList = DataPlaceHolder.Success(result.data))
                    }
                }
                is SharedResult.Error -> {
                    computeAndUpdateState {
                        it.copy(
                            azanList = DataPlaceHolder.Error(
                                message = result.message,
                                actionText = strings.retry
                            )
                        )
                    }
                }
            }
        }
    }

    private fun scheduleNextAzanFetch() {
        fetchNextAzanJob?.cancel()
        fetchNextAzanJob = viewModelScope.launch(Dispatchers.Default) {
            while (true) {
                ensureActive()

                val lastFetchedNextAzan = lastFetchedNextAzan

                @Suppress("ReplaceSafeCallChainWithRun")
                val sharedTimeUntil: SharedTimeUntil? = lastFetchedNextAzan
                    ?.let { dateTimeService.getTimeUntil(it.dateModel, it.timeModel) }
                    ?.takeIf { it.isInFuture }

                if (sharedTimeUntil != null) {
                    updateNextAzanState(
                        azanModel = lastFetchedNextAzan,
                        sharedTimeUntil = sharedTimeUntil
                    )
                    continue
                }

                when (val result = azanService.getNextAzan()) {
                    is SharedResult.Success -> {
                        val azanModel = result.data
                        updateNextAzanState(
                            azanModel = azanModel,
                            sharedTimeUntil = dateTimeService.getTimeUntil(
                                dateModel = azanModel.dateModel,
                                timeModel = azanModel.timeModel
                            )
                        )
                    }
                    is SharedResult.Error ->
                        withContext(Dispatchers.Main) {
                            computeAndUpdateState {
                                it.copy(
                                    nextAzan = DataPlaceHolder.Error(
                                        message = result.message,
                                        actionText = strings.retry
                                    )
                                )
                            }
                        }
                }

                delay(500)
            }
        }
    }

    private suspend fun updateNextAzanState(
        azanModel: SharedAzanModel,
        sharedTimeUntil: SharedTimeUntil
    ) {
        withContext(Dispatchers.Main) {
            val nextAzanNameAndTime = azanModel.displayName + " " + azanModel.displayTime
            val timeUntilNextAzan = localizationService.getTimeUntilNextAzan(sharedTimeUntil)

            lastFetchedNextAzan = azanModel

            computeAndUpdateState {

                val updatedAzanList: DataPlaceHolder<List<State.Azan>> =
                    it.azanList.mapSuccess { azanList ->
                        azanList.map { azan ->
                            azan.copy(isNextAzan = azan.model.id == azanModel.id)
                        }
                    }

                it.copy(
                    azanList = updatedAzanList,
                    nextAzan = DataPlaceHolder.Success(
                        State.NextAzan(
                            id = azanModel.id,
                            title = strings.nextPrayer,
                            text = nextAzanNameAndTime,
                            periodText = timeUntilNextAzan,
                            dateModel = azanModel.dateModel,
                        )
                    )
                )
            }
        }
    }

    private fun listenToSettingsChanges() {
        viewModelScope.launch(Dispatchers.Main) {
            settingsService.eventsFlow
                .collect {
                    reloadAzanList()
                }
        }
    }

    //endregion

    //region State

    /**
     * State
     */
    @Stable
    data class State(
        val azanDate: AzanDate = AzanDate(
            date = SharedDateModel(day = 0, month = 0, year = 0),
            westernFormattedDate = "",
            islamicFormattedDate = ""
        ),
        val nextAzan: DataPlaceHolder<NextAzan> = DataPlaceHolder.Loading,
        val azanList: DataPlaceHolder<List<Azan>> = DataPlaceHolder.Loading,
    ) : BaseState {

        /**
         *
         */
        @Stable
        data class AzanDate(
            val date: SharedDateModel,
            val westernFormattedDate: String,
            val islamicFormattedDate: String
        )

        /**
         *
         */
        @Stable
        data class Azan(
            val model: SharedAzanModel,
            val isNextAzan: Boolean,
            val isNotificationEnabled: Boolean,
        )

        /**
         *
         */
        @Stable
        data class NextAzan(
            val id: String,
            val title: String,
            val text: String,
            val periodText: String,
            val dateModel: SharedDateModel,
        )
    }

    //endregion

}

File: iosApp/iosApp/ui/screens/home/HomeViewModel.swift
//
// Created by Abdullah Essa on 24.05.21.
// Copyright (c) 2021 orgName. All rights reserved.
//

import Foundation
import RxSwift
import shared

class HomeViewModel: ObservableObject {

    //region Private Properties

    private let sharedAzanService: SharedAzanService

    private let sharedDateTimeService: SharedDateTimeService

    private let sharedLocalizationService: SharedLocalizationService

    private let sharedSettingsService: SharedSettingsService
    
    private let sharedTrackingService: SharedTrackingService

    private let sideEffectsSubject = PublishSubject<SideEffect>()

    private let disposableBag = CompositeDisposable()

    private var scheduleNextAzanTimeUpdateDisposeKey: CompositeDisposable.DisposeKey?

    private var loadAzanTimesDisposeKey: CompositeDisposable.DisposeKey?

    //endregion

    //region Public Properties

    @Published var state: HomeViewState = HomeViewState()

    lazy var sideEffects: Observable<SideEffect> = sideEffectsSubject

    var strings: SharedStrings {
        sharedLocalizationService.strings
    }

    //endregion

    //region Init & Deinit

    init(
            sharedAzanService: SharedAzanService,
            sharedDateTimeService: SharedDateTimeService,
            sharedLocalizationService: SharedLocalizationService,
            sharedSettingsService: SharedSettingsService,
            sharedTrackingService: SharedTrackingService
    ) {
        self.sharedAzanService = sharedAzanService
        self.sharedDateTimeService = sharedDateTimeService
        self.sharedLocalizationService = sharedLocalizationService
        self.sharedSettingsService = sharedSettingsService
        self.sharedTrackingService = sharedTrackingService
        sharedTrackingService.trackScreen(screen: Screen.home)
        listenToSettingsChanges()
        scheduleNextAzanTimeUpdate()
        loadAzanTimes(date: sharedDateTimeService.getToday()) // SharedDateModel(day: 1, month: 1, year: 2022)
    }

    private func listenToSettingsChanges() {
        sharedSettingsService.eventsFlow.collect(collector: SwiftFlowCollector<SharedSettingsEvent> { [self] event in
            switch (event) {
            case is SharedNotificationsChangedEvent:
                // Update the isEnabled field in the azan list
                let newAzanList: [PlaceHolder<Azan>] = state.azanList.map { (holder: PlaceHolder<Azan>) in
                    holder.mapSuccess { (item: Azan) in
                        item.copy(isEnabled: sharedSettingsService.isAzanEnabled(azanType: item.azanType))
                    }
                }
                state = state.copy(azanList: newAzanList)
            case is SharedLanguageChangedEvent:
                // Reload all the azan times for the current date to get the new localized strings from shared
                state = state.copy(nextAzan: .loading)
                loadAzanTimes(date: state.azanDate.date)
            default:
                ()
            }
        }
        ) { (unit, error) in
            // code which is executed if the Flow object completed
        }
    }

    deinit {
        disposableBag.dispose()
    }

    //endregion

    //region Public Methods

    func goToNextAzanDay() {
        switch (state.nextAzan) {
        case .success(let value):
            if (value.azan.date == state.azanDate.date) {
                return
            }
            loadAzanTimes(date: value.azan.date)
        default:
            return
        }
    }

    func goToDayBefore() {
        let newDay = sharedDateTimeService.getDayBefore(dateModel: state.azanDate.date)
        loadAzanTimes(date: newDay)
    }

    func goToDayAfter() {
        let newDay = sharedDateTimeService.getDayAfter(dateModel: state.azanDate.date)
        loadAzanTimes(date: newDay)
    }

    func reload() {
        loadAzanTimes(date: state.azanDate.date)
    }
    
    func resetNextAzan() {
        updateState { state in
            state.copy(
                nextAzan: PlaceHolder.loading
            )
        }
    }


    //endregion

    //region Main Logic

    private func scheduleNextAzanTimeUpdate() {
        if let key = scheduleNextAzanTimeUpdateDisposeKey {
            disposableBag.remove(for: key)
        }

        // Schedule next azan interval
        scheduleNextAzanTimeUpdateDisposeKey = disposableBag.insert(
                Observable<Int>.interval(RxTimeInterval.seconds(1), scheduler: MainScheduler.instance)
                        .flatMap { _ in
                            self.getNextAzanSingle().asCompletable().catch{ _ in Completable.empty() }
                        }
                        .subscribe()
        )
    }

    private func loadAzanTimes(date: SharedDateModel) {
        if let key = loadAzanTimesDisposeKey {
            disposableBag.remove(for: key)
        }

        updateState { state in
            state.copy(
                    azanDate: AzanDate(
                            date: date,
                            westernFormattedDate: sharedDateTimeService.getWesternFormattedDate(dateModel: date),
                            islamicFormattedDate: sharedDateTimeService.getIslamicFormattedDate(dateModel: date)
                    ),
                    azanList: HomeViewState.getDummyLoadingList()
            )
        }

        loadAzanTimesDisposeKey = disposableBag.insert(getAzanListSingle(date: date).subscribe())
    }

    //endregion

    //region RXSingles

    private func getNextAzanSingle() -> Single<NextAzan> {
        Single<NextAzan>.create(subscribe: { [self]  emitter in

            // Calculate remaining time
            if let oldNextAzanTime = state.nextAzan.getDataOrNull() {
                let dateModel = oldNextAzanTime.azan.date
                let timeModel = oldNextAzanTime.azan.time
                let timeUntil = sharedDateTimeService.getTimeUntil(dateModel: dateModel, timeModel: timeModel)
                if (timeUntil.isInFuture) {
                    let timeUntil = sharedDateTimeService.getTimeUntil(dateModel: dateModel, timeModel: timeModel)
                    let periodText = sharedLocalizationService.getTimeUntilNextAzan(timeUntil: timeUntil)
                    emitter(.success(NextAzan(
                            azan: oldNextAzanTime.azan,
                            periodText: periodText
                    )))
                    return Disposables.create()
                }
            }

            // Load next azan
            sharedAzanService.getNextAzan(completionHandler: { optionalData, optionalErr in
                if let result: Any = optionalData {
                    switch result {
                    case let success as SharedResultSuccess<SharedAzanModel>:
                        // Success
                        let nextAzanTime = toAzanData(model: success.data, isNext: true, isEnabled: false)
                        let timeUntil = sharedDateTimeService.getTimeUntil(dateModel: nextAzanTime.date, timeModel: nextAzanTime.time)
                        let periodText = sharedLocalizationService.getTimeUntilNextAzan(timeUntil: timeUntil)
                        emitter(.success(NextAzan(azan: nextAzanTime, periodText: periodText)))
                    case let error as SharedResultError:
                        // Error
                        emitter(.failure(error.toNSError()))
                    default:
                        ()
                    }
                }
            })
            return Disposables.create()
        }).do(onSuccess: { [self] (nextAzan: NextAzan) in
            // Update State
            updateState { state in
                let newAzanList: [PlaceHolder<Azan>] = self.state.azanList.map { (holder: PlaceHolder<Azan>) in
                    holder.mapSuccess { (item: Azan) in
                        item.copy(isNext: item.id == nextAzan.azan.id)
                    }
                }
                return state.copy(
                        nextAzan: .success(data: nextAzan),
                        azanList: newAzanList
                )
            }
        }).do(onError: { [self] (error: Error) in
            // Update State
            updateState { state in
                return state.copy(nextAzan: .error(message: (error as NSError).domain))
            }
        })
    }

    private func getAzanListSingle(date: SharedDateModel) -> Single<[SharedAzanModel]> {
        let dataSingle: Single<[SharedAzanModel]> = Single<[SharedAzanModel]>.create(subscribe: { [self]  emitter in
            sharedAzanService.getAzanList(dateModel: date, completionHandler: { optionalData, optionalErr in
                if let result: Any = optionalData {
                    switch result {
                    case let success as SharedResultSuccess<AnyObject>:
                        // Success
                        let data = success.data as! [SharedAzanModel]
                        emitter(.success(data))
                    case let error as SharedResultError:
                        // Error
                        emitter(.failure(error.toNSError()))
                    default:
                        ()
                    }
                }
            })
            return Disposables.create()
        })

        let delaySingle: Single<Bool> = Observable.just(true)
                .delay(.milliseconds(300), scheduler: MainScheduler.instance)
                .asSingle()

        return Single.zip(dataSingle, delaySingle).map({ (data: [SharedAzanModel], _: Bool) in data })
                .do(onSuccess: { [self] (azanTimes: [SharedAzanModel]) in
                    // Update State
                    let azanData = azanTimes.map({ (item: SharedAzanModel) -> PlaceHolder<Azan> in
                        let isEnabled = sharedSettingsService.isAzanEnabled(azanType: item.azanType)
                        let isNextAzan: Bool = item.id == state.nextAzan.getDataOrNull()?.azan.id
                        return .success(data: toAzanData(model: item, isNext: isNextAzan, isEnabled: isEnabled))
                    })
                    updateState { state in
                        state.copy(azanList: azanData)
                    }
                })
                .do(onError: { [self] (error: Error) in
                    // Update State
                    updateState { state in
                        return state.copy(azanList: HomeViewState.getDummyErrorList(message: (error as NSError).domain))
                    }
                })
    }

    //endregion

    //region Helpers

    private func updateState(_ updateBlock: (HomeViewState) -> HomeViewState) {
        state = updateBlock(state)
    }

    private func toAzanData(
            model: SharedAzanModel,
            isNext: Bool,
            isEnabled: Bool
    ) -> Azan {
        Azan(
                id: model.id,
                displayName: model.displayName,
                displayTime: model.displayTime,
                azanType: model.azanType,
                date: model.dateModel,
                time: model.timeModel,
                isNext: isNext,
                isEnabled: isEnabled
        )
    }

    //endregion

    //region SideEffects

    class SideEffect {
    }

    //endregion
}
